torchtune/datasets/_iterable_base.py

```diff
@@ -0,0 +1,43 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+from abc import ABC, abstractmethod
+from typing import Any, Iterator
+
+from torch.utils.data import IterableDataset
+
+
+class TuneIterableDataset(IterableDataset, ABC):
```

Comment:
We need this guy to interact with packing and IIUC I don't believe this is currently happening?

The algo we should implement is this:

1. One batch can be made of *multiple* calls to next. We keep taking until we exceed the max seq len. When we do, we put the last one aside (we'll use it to start the next batch), pad the current one to max len and return.
2. The calls to next will go to the *interleaved* dataset, therefore we automatically construct mixed batches from multiple datasets without much effort
3. Also, every time we call next we should make space for logging transforms (which we are, you already wrote them). I think it's ok to make your metrics transforms and aggregators an optional property here so the semantics are clearer